cmake/modules/BuildQAT.cmake

```diff
@@ -0,0 +1,48 @@
+function(build_qat)
+  include(FindMake)
+  find_make("MAKE_EXECUTABLE" "make_cmd")
+
+  set(QAT_BINARY_DIR ${CMAKE_BINARY_DIR}/src/qatlib)
+  set(QAT_INSTALL_DIR ${QAT_BINARY_DIR}/install)
+  set(QAT_INCLUDE_DIR ${QAT_INSTALL_DIR}/include)
+  set(QAT_LIBRARY_DIR ${QAT_INSTALL_DIR}/lib)
+  set(QAT_LIBRARY ${QAT_LIBRARY_DIR}/libqat.a)
+  set(QAT_USDM_LIBRARY ${QAT_LIBRARY_DIR}/libusdm.a)
+
+  # this include directory won't exist until the install step, but the
+  # imported targets need it early for INTERFACE_INCLUDE_DIRECTORIES
+  file(MAKE_DIRECTORY "${QAT_INCLUDE_DIR}")
+
+  set(configure_cmd env CC=${CMAKE_C_COMPILER} ./configure --prefix=${QAT_INSTALL_DIR})
+  # disable systemd or 'make install' tries to write /usr/lib/systemd/system/qat.service
+  list(APPEND configure_cmd --disable-systemd)
```

Comment:
The option `--disable-systemd` will install qatlib in [standalone mode](https://github.com/intel/qatlib/blob/abe15d7bfc083117bfbb4baee0b49ffcd1c03c5c/INSTALL#L257), so there is no _qat_ service running, which [is the preferred way to run on containers where a subset of VFs are explicitly passed through and so no qatmgr managed resource mgr is needed.](https://github.com/intel/qatlib/blob/abe15d7bfc083117bfbb4baee0b49ffcd1c03c5c/INSTALL#L499C8-L499C8)  However IMHO, if the Ceph cluster is running on bare metal not in container, it maybe better to install qatlib in [managed mode](https://github.com/intel/qatlib/blob/abe15d7bfc083117bfbb4baee0b49ffcd1c03c5c/INSTALL#L366) with option `--enable-service` so that qatmgr manages allocation of the VF resources to each process on process startup from performance perspective. On the other hand, if qatlib packages is installed on Linux distributions such as Fedora 34+, RHEL 9.1+, or CentOS 9 Stream by running command `sudo dnf -y install qatlib-devel`, I think [qatlib is installed in managed mode](https://intel.github.io/quickassist/qatlib/install.html#installing-from-rpm-packages), where _qat_ service is enabled. IMHO, it maybe better to distinguish the installation between managed mode and standalone mode. Although it seems to be able to run mutil-process test to use QAT devices by installing qatlib in standalone mode on bare metal, the performance of qat is worse than its installation in managed mode.

Reply:
oops, thanks. i'd only added --disable-systemd because the 'make install' step doesn't respect the --prefix, and we don't want to be installing things to system directories (or require root) during the ceph build. we could work around that by avoiding the 'make install' step and pointing at artifacts in the source/build directories instead of the install dir

but we'd also need to package the qat service and its systemd stuff to be installed along with ceph, and that gets complicated and risks conflicts if someone installs ceph on a system that also has qat packages installed

it's also tricky to do it one way (standlone) for container builds and another way (managed) for package builds, because afaik our container images are built from the ceph packages

it does make sense for cmake to offer both options - and maybe WITH_SYSTEM_QATLIB=ON means managed and OFF means standalone - but we may not be able to support both for upstream users

our hands are further tied by the fact that debian doesn't have system packages, so the WITH_SYSTEM_QATLIB=OFF build would be the only thing we could ship atm

Reply:
@liumiaomiaoIntel @mkogan1 @mattbenjamin is everyone okay with this decision to only support the standalone mode for ceph releases (both packages and containers)?

> In standalone mode, there is no qat service running. This is the preferred way to run on containers where a subset of VFs are explicitly passed through and so no qatmgr managed resource mgr is needed.

on the container side, do we need any additional support from cephadm to accomplish this part?

Reply:
Intel QAT devices delivers turn-key standard PCI Express access to hardware acceleration and their resources need to be managed to allocate QAT de/compression and cryption VFs to user's processes. In bare metal, qat service installed by the managed mode is used to manage QAT resources. In container, QAT VF resources also need to be managed, for example [QAT device plugin](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/cmd/qat_plugin/README.md) who manages QAT resources in Kubernetes. In this way, cephadm might need to manage QAT VFs and path through a subset of VFs to each OSD containers without conflict.

Reply:
@liumiaomiaoIntel that's very helpful, thank you